File: python/bishop/bishop.py
from point import Point

class Bishop:
    def __init__(self, source, dest):
        self.source = Point(source)
        self.dest= Point(dest)

    def moves(self):
        if self.source == self.dest:
            return 0
        elif not self.validMove():
            return -1
        return 1 if self.sameDiagonal() else 2

    def validMove(self):
        return (self.source.x + self.source.y) % 2 == (self.dest.x + self.dest.y) % 2

    def sameDiagonal(self):
        return abs(self.dest.x - self.source.x) == abs(self.dest.y - self.source.y)

File: ruby/lib/bishop.rb
require 'point'

class Bishop
  def initialize(source, destination)
    @source = Point.new(source)
    @destination = Point.new(destination)
  end

  def valid_move
    (@source.x + @source.y) % 2 === (@destination.x + @destination.y) % 2
  end

  def moves
    return 0 if @source === @destination
    return -1 unless valid_move
    return 1 if same_diagonal
    2
  end

  def same_diagonal
    return (@destination.x - @source.x).abs == (@destination.y - @source.y).abs
  end
end
